src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala

```diff
@@ -37,6 +40,31 @@ class CampaignResource(extractAuth: Directive1[AuthedNamespaceScope],
     } yield campaignId
   }
 
+  /**
+    * Create and immediately launch a retry-campaign for all the devices that were processed by `mainCampaign` and
+    * failed with a code `failureCode`. We assume there is at least one failed device. Hence, if there were not any
+    * such devices in `mainCampaign`, throw an error.
+    */
+  def retryFailedDevices(ns: Namespace, mainCampaign: Campaign, request: RetryFailedDevices): Future[Unit] =
+    campaigns.fetchFailedDevices(mainCampaign.id, request.failureCode).map(_.toSeq).flatMap {
+      case Nil =>
```

Comment:
Instead of converting the list of devices to and from `Seq`, I'd suggest to simply replace `case Nil` with `case devicesIds if devicesIds.isEmpty`. While it decreases readability a bit, it will:
1. Avoid unnecessary and potentially costly conversion
2. Remove non-explicit dependency on the standard Scala library implementation of `Set` being `List`. They might change it anytime to `Vector` and then your code will stop working because `Nil` won't ever match `Vector`.

Reply:
Yeah I wasn't too happy about that myself. Thanks!